README.md
# MMM-FreeBox-Monitor

[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg)](https://github.com/tataille/MMM-FreeBox-Monitor/graphs/commit-activity)
[![License](https://img.shields.io/badge/License-Apache%202.0-blue.svg)](https://opensource.org/licenses/Apache-2.0)
[![Issues](https://img.shields.io/github/issues/tataille/MMM-FreeBox-Monitor.svg?style=flat-square)](https://github.com/tataille/MMM-FreeBox-Monitor/issues)
[![NPM](https://img.shields.io/npm/dm/magic-mirror-module-freeboxmonitor)](https://www.npmjs.com/package/magic-mirror-module-freeboxmonitor)

<a href="https://www.buymeacoffee.com/jeanmarctaz"><img src="https://img.buymeacoffee.com/button-api/?text=Buy me a beer&emoji=🍺&slug=jeanmarctaz&button_colour=40DCA5&font_colour=ffffff&font_family=Cookie&outline_colour=000000&coffee_colour=FFDD00" /></a>

## MagicMirror Module for Freebox Delta

![MMM-FreeBox-Monitor: The module for MagicMirror. ](https://github.com/tataille/MMM-FreeBox-Monitor/blob/master/2016-06-30_14-34-46.png)

[Freebox API ](http://mafreebox.freebox.fr/doc/index.html)


This module is intended to be used in [MagicMirror²!](https://github.com/MichMich/MagicMirror).

The `MMM-FreeBox-Monitor` displays data from your FreeBox v6 server (Revolution) from the french ISP free. It can combine multiple data

## Using the module

To use this module, add it to the modules array in the `config/config.js` file:
````javascript
modules: [
	{
		module: 'MMM-FreeBox-Monitor',
		position: 'top_left',	// This can be any of the regions. Best results in left or right regions.
		config: {
			mirrorName: "My Magic Mirror",
			ip: "http://XXX.XXX.XXX.XXX"
			// See 'Configuration options' for more information.
		}
	}
]
````

## Configuration options

The following properties can be configured:

| **Option**| **Description** |
| --- | --- |
| `ip` | The IP address of the FreeBox Server. You can use the internal or external IP address of the FreeBox Server, you can use [http://www.myipaddress.com/](http://www.myipaddress.com/) service to get it. Default FreeBox Server ip is [http://192.168.0.254](http://192.168.0.254).<br><br>**Possible values:** any IP address prefixed by _http://_ |
| `maxCallEntries` | The maximum number of missed calls shown.<br><br>**Possible values:** ``0`` - ``100`` <br>**Default value:** ``3`` |
| `displayMissedCalls` | Display missed calls table.<br/><br/> **Possible values:** ``true`` or ``false``<br/>**Default value:** ``true`` |
| `displayDownloads` | Display downloaded files table.<br/><br/> **Possible values:** ``true`` or ``false``<br/>**Default value:** ``true`` |
| `displaySystemData` | Display System Data.<br/><br/> **Possible values:** ``true`` or ``false``<br/>**Default value:** ``true`` |
| `displayFirmware` | Display OS Firmware.<br/><br/> **Possible values:** ``true`` or ``false``<br/>**Default value:** ``false`` |
| `displayConnectedDevices` | Display Total number of connected devices by Wifi Access Point type.<br/><br/> **Possible values:** ``true`` or ``false``<br/>**Default value:** ``false`` |
| `mirrorName` | The mirror name used to identify the mirror on Freebox Server. Useful in case of multiple mirrors. <br/><br/> **Possible values:** ``string``<br/>**Default value:** ``My Magic Mirror`` |
| `requestRefresh` | The refresh interval in seconds to request refrest on Freebox Server. <br/><br/> **Possible values:** ``30`` \- ``300``<br/>**Default value:** ``30`` |

#### FreeBox Monitor registration:

Before starting the Magic Mirror. Open your Freebox admin page at [http://mafreebox.freebox.fr](http://mafreebox.freebox.fr) -> _Paramètres de la Freebox_ -> _Gestion des accès_. Check that _Permettre les nouvelles demandes d'association_ is checked.

![MMM-FreeBox-Monitor: Application association ](https://github.com/tataille/MMM-FreeBox-Monitor/blob/master/2016-11-23_21-43-40.png)

The first time the Magic Mirror starts, a request is sent to Freebox Server to register the application. Select ``Oui`` on LCD display to register the mirror.

![MMM-FreeBox-Monitor: Registration ](https://github.com/tataille/MMM-FreeBox-Monitor/blob/master/20160630_234117.png)

Once session is opened, a file ``freebox.txt`` is created at the root of Magic Mirror installation directory. If you change the configuration for ``mirrorName`` or ``ip``, you must delete this file.

## Dependencies
- installed via `npm install`


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.2] - 2023-06-29

### Major changes

- CHANGELOG.md Added
- Version of API now reflect the deployed API version from Freebox
- Tested on Freebox Delta & Revolution box
- Docker-compose file added to allow to start a local instance of MagicMirror
- Change date and caller Id emplacement
- Refactor Download section display
- Download section will disappear if no download is in progress or done
- Add Status section
- Add Firmware version display (optional)

### Changed features

- Change the upload / download rate display to reflect the correct Freebox connection rates
- Update `eslint` from `~7.18.0` to `^8.13.0`

### Fixed

- Application name was __undefined__ at permission prompt

## [1.0.1] - 2022-05-02

First release of MMM-FreeBox-Monitor module


freebox.js
/* global Module */

/* Magic Mirror
 * Module: freebox
 *
 * By Jean-Marc Taillant
 * MIT Licensed.
 */

 
 
Module.register("freebox",{

	// Default module config.
	defaults: {
		fade: true,
		fadePoint: 0.25,
		language: "fr",
		maxCallEntry: 3,
		mirorName: "myMirror",
		requestRefresh: 30
	},
	
	
	start: function() {
		Log.info("Starting module: " + this.name);
		this.errorMessage = "";
		this.callsTable = [];
		this.downloadsTable = "";
		this.needRefresh = true;
		this.wrapper = "";
		var self = this;
			setInterval(function() {
			self.updateDom(); // no speed defined, so it updates instantly.
		}, 5000); 
		moment.locale(config.language);		
		this.sendSocketNotification("CONNECT", {
				config: this.config
		});
		


	},
	// Define required scripts.
	getScripts: function() {
		return ["moment.js","progressbar.js"];
	},
	// Define required scripts.
	getStyles: function() {
		return ["freebox.css"];
	},
	// Override dom generator.
	getDom: function() {
		if (this.needRefresh){
		/*var wrapper = document.createElement("div");
		wrapper.id = "msg";
		wrapper.innerHTML = "Connection..";*/
		this.wrapper = document.createElement("div");
		var statusWrapper = document.createElement("div");
		statusWrapper.id = "status";
		var callsWrapper = document.createElement("div");
		callsWrapper.id = "calls";
		var callsHeaderWrapper = document.createElement("div");
		callsHeaderWrapper.id = "callsheader";		
		
		
		var downloadsWrapper = document.createElement("div");
		downloadsWrapper.id = "downloads";
		var downloadsHeaderWrapper = document.createElement("div");
		downloadsHeaderWrapper.id = "downloadsHeader";		
	
		// Style Wrappers
		statusWrapper.className = "date normal medium";
		callsWrapper.className = "date normal medium";
		downloadsWrapper.className = "time bright large light";
		this.wrapper.appendChild(statusWrapper);
		callsWrapper.appendChild(callsHeaderWrapper);
		this.wrapper.appendChild(callsWrapper);
		this.wrapper.appendChild(downloadsWrapper);
		downloadsWrapper.appendChild(downloadsHeaderWrapper);
		
		//Where the magic happends
		if (this.errorMessage != ""){
			statusWrapper.innerHTML = this.errorMessage;
		}else {
			if (this.callsTable.length > 0){
				callsHeaderWrapper.innerHTML = "Appels Manqués";
				callsHeaderWrapper.className = "tableheader align-left";
				/*var calls = document.getElementById("calls");				
				var child = document.getElementById("callTable"); 
				if ( child !== null)
					calls.removeChild(child);*/
				var table = document.createElement("table");
				table.id= "callsTable";
				table.className = "small";
				if ( this.config.maxCallEntry> this.callsTable.length)
					 this.config.maxCallEntry = this.callsTable.length;				
				for (var mc in this.callsTable) {					
					var missedCall = this.callsTable[mc];
					var row = document.createElement("tr");
					table.appendChild(row);
					var dateCell = document.createElement("td");
					dateCell.className = "day";
					var date = moment(missedCall.datetime, "X").format("ddd DD HH:mm");
					dateCell.innerHTML = date;
					row.appendChild(dateCell);
					
					var callerCell = document.createElement("td");
					callerCell.innerHTML = missedCall.name;
					callerCell.className = "align-right bright max-temp";
					row.appendChild(callerCell);

					if (this.config.fade && this.config.fadePoint < 1) {
						if (this.config.fadePoint < 0) {
							this.config.fadePoint = 0;
						}
						var startingPoint = this.callsTable.length * this.config.fadePoint;
						var steps = this.callsTable.length - startingPoint;
						if (mc >= startingPoint) {
							var currentStep = mc - startingPoint;
							row.style.opacity = 1 - (1 / steps * currentStep);
						}
					} 
				}
				callsWrapper.appendChild(table);				
			}
			if ( this.downloadsTable != "" ){
				var cpt=0;
				downloadsHeaderWrapper.innerHTML = "Téléchargements";
				downloadsHeaderWrapper.className = "tableheader align-left";
				var desc = document.createElement("table");
				desc.id= "downloadTable";
				desc.className = "small";
		
				for (var d in this.downloadsTable.done) {
					var fileDone = this.downloadsTable.done[d];
					var row = document.createElement("tr");
					desc.appendChild(row);
					var pourcentCell = document.createElement("td");
					var pcontainer = document.createElement("div");
					pcontainer.id="container_"+(cpt++);
					pcontainer.className = "moviepercent";
					pourcentCell.appendChild(pcontainer);
					
					var bar = new ProgressBar.Circle(pcontainer, {
						color: '#aaa',
						// This has to be the same size as the maximum width to
						// prevent clipping
						strokeWidth: 4,
						trailWidth: 1,
						easing: 'easeInOut',
						duration: 2500,
						text: {
							autoStyleContainer: false
						},
						from: { color: '#aaa', width: 1 },
						to: { color: '#333', width: 4 },
						// Set default step function for all animate calls
						step: function(state, circle) {
						circle.path.setAttribute('stroke', state.color);
						circle.path.setAttribute('stroke-width', state.width);

						var value = Math.round(fileDone.rx_pct / 10000 * 100);
						if (value === 0) {
							circle.setText('');
						} else {
							circle.setText(value);
						}

						}
					});
					row.appendChild(pourcentCell);
					
					bar.text.style.fontFamily = '"Raleway", Helvetica, sans-serif';
					bar.text.style.fontSize = '10px';
					//bar.animate(Math.round(fileDone.rx_pct / 10000));
					//pourcentCell.innerHTML = Math.round(fileDone.rx_pct / 100);
					//pourcentCell.className = "align-right bright max-temp";
					
					var movieCell = document.createElement("td");
					movieCell.className = "moviename";					
					movieCell.innerHTML = fileDone.name;
					row.appendChild(movieCell);	

					var yearCell = document.createElement("td");
					yearCell.className = "day";					
					yearCell.innerHTML = fileDone.year;
					row.appendChild(yearCell);						
				}
				for (var d in this.downloadsTable.seeding) {
					var fileDone = this.downloadsTable.seeding[d];
					var row = document.createElement("tr");
					desc.appendChild(row);
					var pourcentCell = document.createElement("td");

					var pcontainer = document.createElement("div");
					pcontainer.id="container_"+(cpt++);
					pcontainer.className = "moviepercent";
					pourcentCell.appendChild(pcontainer);
					
					var bar = new ProgressBar.Circle(pcontainer, {
						color: '#aaa',
						// This has to be the same size as the maximum width to
						// prevent clipping
						strokeWidth: 4,
						trailWidth: 1,
						easing: 'easeInOut',
						duration: 2500,
						text: {
							autoStyleContainer: false
						},
						from: { color: '#aaa', width: 1 },
						to: { color: '#333', width: 4 },
						// Set default step function for all animate calls
						step: function(state, circle) {
						circle.path.setAttribute('stroke', state.color);
						circle.path.setAttribute('stroke-width', state.width);

						var value = Math.round(fileDone.rx_pct / 10000 * 100);
						if (value === 0) {
							circle.setText('');
						} else {
							circle.setText(value);
						}

						}
					});
					row.appendChild(pourcentCell);
					bar.text.style.fontFamily = '"Raleway", Helvetica, sans-serif';
					bar.text.style.fontSize = '10px';
					//bar.animate(Math.round(fileDone.rx_pct / 10000));

					
					
					var movieCell = document.createElement("td");
					movieCell.className = "moviename";					
					movieCell.innerHTML = fileDone.name;
					row.appendChild(movieCell);	

					var yearCell = document.createElement("td");
					yearCell.className = "day";					
					yearCell.innerHTML = fileDone.year;
					row.appendChild(yearCell);								
				}
				//
				for (var d in this.downloadsTable.downloading) {
					var fileDone = this.downloadsTable.downloading[d];
					var row = document.createElement("tr");
					desc.appendChild(row);
					var pourcentCell = document.createElement("td");

					var pcontainer = document.createElement("div");
					pcontainer.id="container";
					pcontainer.className = "moviepercent";
					pourcentCell.appendChild(pcontainer);
					
					var bar = new ProgressBar.Circle(pcontainer, {
						color: '#aaa',
						// This has to be the same size as the maximum width to
						// prevent clipping
						strokeWidth: 4,
						trailWidth: 1,
						easing: 'easeInOut',
						duration: 2500,
						text: {
							autoStyleContainer: false
						},
						from: { color: '#aaa', width: 1 },
						to: { color: '#333', width: 4 },
						// Set default step function for all animate calls
						step: function(state, circle) {
						circle.path.setAttribute('stroke', state.color);
						circle.path.setAttribute('stroke-width', state.width);

						var value = Math.round(fileDone.rx_pct / 10000 * 100);
						if (value === 0) {
							circle.setText('');
						} else {
							circle.setText(value);
						}

						}
					});
					row.appendChild(pourcentCell);
					bar.text.style.fontFamily = '"Raleway", Helvetica, sans-serif';
					bar.text.style.fontSize = '10px';
					bar.animate(Math.round(fileDone.rx_pct / 10000));
					
					
					
					var movieCell = document.createElement("td");
					movieCell.className = "moviename";					
					movieCell.innerHTML = fileDone.name;
					row.appendChild(movieCell);	

					var yearCell = document.createElement("td");
					yearCell.className = "day";					
					yearCell.innerHTML = fileDone.year;
					row.appendChild(yearCell);								
				}
				
				//
				downloadsWrapper.appendChild(desc);
				this.needRefresh = false;
			}
		}
		}
		
		return this.wrapper;
		
	},
	// Override socket notification handler.
	socketNotificationReceived: function(notification, payload) {		
		var self = this;				
		if (notification === "FREEBOX_ERROR") {
			self.callsTable = [];	
			self.downloadsTable ="";
			self.errorMessage = payload.error;			
		}else if (notification == "FREEBOX_MSG"){
			self.errorMessage = "";
			if ( payload.type === "downloads"){					
				self.downloadsTable ="";
				self.downloadsTable =JSON.parse(JSON.stringify(payload.value));				
			}else if ( payload.type === "calls"){			
				self.callsTable = [];	
				var newArray = payload.value.slice(0, this.config.maxCallEntry);
				for (var mc in newArray) {
					self.callsTable.push(newArray[mc]);
				}		
			}else if ( payload.type === "connection"){				
				if (payload.value === "Connected"){
					self.getCalls();
					self.getDownloads();
					setInterval(function() {self.getCalls();self.getDownloads();	}, (this.config.requestRefresh) * 1000);					
					//setInterval(function() {self.getCalls();self.getDownloads();	}, 10000);					
				}
				
			}		
		}
		self.needRefresh = true;
	},
	getCalls: function(){
		this.sendSocketNotification("GET_CALLS", {
				config: this.config
		});
			
	},
	getDownloads: function(){
		this.sendSocketNotification("GET_DOWNLOADS", {
				config: this.config
		});
			
	}

});


node_helper.js
/* Magic Mirror
 * Node Helper: Newsfeed
 *
 * By Michael Teeuw http://michaelteeuw.nl
 * MIT Licensed.
 */

var NodeHelper = require("node_helper")
var fs = require("fs");
var Freeboxapi = require("./freeboxapi.js");
var fapi;


module.exports = NodeHelper.create({
	// Subclass start method.
	start: function() {
		var self = this;
		self.fapi = new Freeboxapi("..");
		self.fapi.onError(function(fapi, error) {
				self.sendSocketNotification("FREEBOX_ERROR", {
					error: error
				});
			});
		self.fapi.onMsg(function(fapi, msg) {
				self.sendSocketNotification("FREEBOX_MSG", {
					value: msg.value,
					type: msg.type
				});
			});
		console.log("Starting module: " + this.name);
		
	},

	// Subclass socketNotificationReceived received.
	socketNotificationReceived: function(notification, payload) {
		var self = this;
		console.log("FREEBOX: "+JSON.stringify(notification)+ " " +payload);
		if (notification === "GET_CALLS") {
			self.fapi.getCalls();
			return;
		}else if (notification === "GET_DOWNLOADS") {
			self.fapi.getDownloads();
			return;
		}else if (notification === "GET_CONNECTED_DEVICES") {
			self.fapi.getConnectedDevices();
			return;
		}else if (notification === "GET_CONNECTIONSTATUS") {
			self.fapi.getConnectionStatus();
			return;
		}else if (notification === "GET_SYSTEMINFO") {
			self.fapi.getSystemInfo();
			return;
		}else if (notification === "CONNECT"){
			self.fapi.configure(payload.config);
			self.fapi.readFreeboxVersion();
			return;
		} 
			
	}
});


mounts/config/config.js
/* MagicMirror² Config
 *
 * By Jean Marc TAILLANT
 * MIT Licensed.
 *
 * For more information on how you can configure this file
 * see https://docs.magicmirror.builders/configuration/introduction.html
 * and https://docs.magicmirror.builders/modules/configuration.html
 *
 * You can use environment variables using a `config.js.template` file instead of `config.js`
 * which will be converted to `config.js` while starting. For more information
 * see https://docs.magicmirror.builders/configuration/introduction.html#enviromnent-variables
 */
let config = {
	address: "0.0.0.0",	// Address to listen on, can be:
							// - "localhost", "127.0.0.1", "::1" to listen on loopback interface
							// - another specific IPv4/6 to listen on a specific interface
							// - "0.0.0.0", "::" to listen on any interface
							// Default, when address config is left out or empty, is "localhost"
	port: 8080,
	basePath: "/",			// The URL path where MagicMirror² is hosted. If you are using a Reverse proxy
					  		// you must set the sub path here. basePath must end with a /
	ipWhitelist: [],	// Set [] to allow all IP addresses
															// or add a specific IPv4 of 192.168.1.5 :
															// ["127.0.0.1", "::ffff:127.0.0.1", "::1", "::ffff:192.168.1.5"],
															// or IPv4 range of 192.168.3.0 --> 192.168.3.15 use CIDR format :
															// ["127.0.0.1", "::ffff:127.0.0.1", "::1", "::ffff:192.168.3.0/28"],

	useHttps: false, 		// Support HTTPS or not, default "false" will use HTTP
	httpsPrivateKey: "", 	// HTTPS private key path, only require when useHttps is true
	httpsCertificate: "", 	// HTTPS Certificate path, only require when useHttps is true

	language: "fr",
	locale: "fr-FR",
	logLevel: ["INFO", "LOG", "WARN", "ERROR","DEBUG"],
	timeFormat: 24,
	units: "metric",

	modules: [
		{
			module: 'MMM-FreeBox-Monitor',
			header: "FREEBOX",
			position: 'top_right',	// This can be any of the regions. Best results in left or right regions.
			config: {
				mirrorName: "My Magic Mirror",
				ip: "http://mafreebox.freebox.fr",
				maxCallEntries: 7,
				requestRefresh: 10,
				displaySystemData: true,
				displayFirmware: true,
				// See 'Configuration options' for more information.
			}
		},
		{
			module: "clock",
			position: "top_left"
		},
		{
			module: "calendar",
			a: "US Holidays",
			position: "top_left",
			config: {
				calendars: [
					{
						symbol: "calendar-check",
						url: "https://www.data.gouv.fr/fr/datasets/r/15366b3e-70ec-4fe5-8ed4-2cd6cf28b38b"
					}
				]
			}
		},
		{
			module: "compliments",
			position: "lower_third",
			config: {
				compliments: {
					anytime: [
						"Bonjour!"
					],
				}
			}
		},
		{
			module: "helloworld",
			position: "bottom_bar", // This can be any of the regions.
			config: {
			  // See 'Configuration options' for more information.
			  text: "Neque porro quisquam est qui dolorem ipsum quia dolor sit amet, consectetur, adipisci velit",
			},
		  },
	]
};

/*************** DO NOT EDIT THE LINE BELOW ***************/
if (typeof module !== "undefined") {module.exports = config;}


freeboxapi.js
/**
 * Created by taillan2 on 24/06/2016.
*/

var reqjson = require('request-json');
var ptn = require('parse-torrent-name');
var imdb = require('imdb-api');
var omdb = require('omdb');
var CryptoJS = require("crypto-js");
var fs = require("fs");
var _ = require("underscore");

var Freeboxapi = function(ip) {
	var version = "";
	var mirrorName = "";
	var apsMap = new Map();
	var nameSize = 10;
	var apsWithDevices = new Map();
	var self = this;
	var config = {
		track_id: "",
		app_token: "",
		appId: "",
		challenge: "",
		session_token: ""
	};

    this.configure =  function(data){
        ip = data.ip;
		mirrorName = data.mirrorName;
    };

	this.readFreeboxVersion = function(){
		var client = reqjson.createClient(ip);
        client.get('/api_version', callbackVersion);
	}

	var callbackVersion = function(error, response, body){
		if (!error) {
            var info = JSON.parse(JSON.stringify(body));            
            config.version = info.api_version.split('.')[0];
			console.debug('API Version is: '+config.version);
			self.authorize("fr.freebox",'MM Freebox Monitor',"1.0.2",mirrorName);
        }
        else {
            console.error('VERSION - Error happened: ' + error);
        }
	}

	var callbackSystemInfo = function(error, response, body){
		if (!error) {
            var info = JSON.parse(JSON.stringify(body));    
			console.debug(info);                    
			if (info.success === true){				
				msg = {
					value:  info.result,
					type: "systemInfo"
				};
				fetchCallback(self, msg);
			}else{
				config.challenge = info.result.challenge;
                openSession(callbackOpenSession);
			}
        }
        else {
            console.error('System Info - Error happened: ' + error);
        }
	}

	var authCallback = function(error, response, body){
        if (!error) {
            var info = JSON.parse(JSON.stringify(body));
			console.debug(info);            
            config.app_token = info.result.app_token;
            config.track_id = info.result.track_id;
			checkAuthorize(config.track_id, callbackCheckAuthorize);
        }
        else {
            console.error('Error happened: ' + error);
        }
    };

	var connectionStatusCallback = function(error, response, body){
        if (!error) {
            var info = JSON.parse(JSON.stringify(body));
			console.debug(info);
        	if (info.success === true){				
				msg = {
					value:  info.result,
					type: "connectionStatus"
				};
				fetchCallback(self, msg);
			}else{
				console.info(info.result);
				config.challenge = info.result.challenge;
                openSession(callbackOpenSession);
			}
        }
        else {
            console.error('Error happened: ' + error);
        }
    };


    var callbackCalls = function(error, response, body){
        if (!error) {
			var info = JSON.parse(JSON.stringify(body));
			if (info.success === true){
            			var info = JSON.parse(JSON.stringify(body));				
            			var calls = info.result;
            			var today = new Date();
            			var start = new Date();
            			start.setDate(start.getDate() - 10);
            			start.setHours(0,0,0,0);
            			start = start.getTime()/1000;
				var end = new Date();
				end.setDate(end.getDate() - 30);
				end.setHours(23,59,59,00);
				end = end.getTime()/1000;            			
				var filtered = _.where(calls, {type: "missed"});
				value = "";
				if ( filtered.length > 1){
					value = filtered.length + " missed calls";
				}else
					value = filtered.length + " missed call";
				msg = {
					value:  filtered,
					type: "calls"
				};
				fetchCallback(self, msg);
			}else{
				config.challenge = info.result.challenge;
                		openSession(callbackOpenSession);
			}
        }
        else {
            console.error('Error happened: ' + error);
        }
    };

	var callbackDownloads = function(error, response, body){
        if (!error) {
            var info = JSON.parse(JSON.stringify(body));	
			console.debug(info);		
            var downloads = info.result;
            var done = _.where(downloads, {status: "done"});
			var seeding = _.where(downloads, {status: "seeding"});
			var downloading = _.where(downloads, {status: "downloading"});
			done.forEach( function( item ) {
				var mov = ptn(item.name);
				item.name = trunckName(mov.title, nameSize);
			});
			seeding.forEach( function( item ) {
				var mov = ptn(item.name);
				item.name = trunckName(mov.title, nameSize);
			});
			downloading.forEach( function( item ) {
				var mov = ptn(item.name);
				item.name = trunckName(mov.title, nameSize);
			});
			files = {
				done: done,
				seeding: seeding,
				downloading: downloading
			};
			msg = {
				value:  files,
				type: "downloads"
			};
			fetchCallback(self, msg);            
            
        }
        else {
            console.error('Error happened: ' + error);
        }
    };

	var trunckName = function (name, maxSize) {
		if (name.length > maxSize) {
		  return name.slice(0, maxSize) + "...";
		} else {
		  return name;
		}
	  }

    var callbackCheckAuthorize = function(error, response, body){		
        if (!error) {
            var info = JSON.parse(JSON.stringify(body));			
			if (info.success === false){
				error = info.msg;
				fetchFailedCallback(self, error);
			}else if (info.result.status =='pending'){
                		setTimeout( function(){checkAuthorize(config.track_id, callbackCheckAuthorize)}, 5000);
            		}else if (info.result.status =='granted'){
				getChallenge(callbackGetChallenge);
            		}else{
				error = "Timeout";
				fetchFailedCallback(self, error);
			}
        }
        else {
            console.error('Error happened: ' + error);
        }
    };

	var callbackGetChallenge = function(error, response, body){
        if (!error) {
            var info = JSON.parse(JSON.stringify(body));
			console.debug("CallbackGetChallenge -> "+ info);            
            if (info.success == true){
                config.challenge = info.result.challenge;
                setTimeout( function(){openSession(callbackOpenSession)}, 2000);
            }

        }
        else {
            console.error('Error happened: ' + error);
        }
    };

	var apListCallback = function(error, response, body){
		if (!error) {
            var info = JSON.parse(JSON.stringify(body));	
			console.debug(info);		
            var apList = info.result;
			for(var i = 0; i < apList.length; i++)
			{
				apsMap.set(apList[i].id, apList[i]);
				apsMap.get(apList[i].id).status = "unknow";

			}
			apsWithDevices = new Map();
			apsMap.forEach(function(ap,ap_id, ) {
				console.debug("Retrieving connected devices for ap "+ap_id+ " "+ ap.name);
				getApDevices(ap_id);
			});
		}
        else {
            console.error('Error happened: ' + error);
        }
	};

	var getApDevices = function(ap_id){
		var client = reqjson.createClient(ip);
        client.headers['X-Fbx-App-Auth'] = config.session_token;
		//retrieve ap list
        client.get('/api/v'+config.version+'/wifi/ap/'+ap_id+'/stations', (error, response, body) => {
			if (!error) {
				var info = JSON.parse(JSON.stringify(body));	
				console.debug(info);		
				var devicesMap = info.result;
				if (devicesMap) {
					apsWithDevices.set(apsMap.get(ap_id).name,  devicesMap.length);
					apsMap.get(ap_id).number = devicesMap.length;
					console.debug( devicesMap.length +" devices connected on ap "+apsMap.get(ap_id).name);
				}
				apsMap.get(ap_id).status = "done";
				var send = true;
				apsMap.forEach(function(ap,id, ) {
					console.log("Checking status for "+id+" "+ap.status);
					if ( ap.status === "unknow"){
						send = false;
					}
				});
				
				if ( send === true) {
					//var devices  = _.where(downloads, {status: "done"});
					 msg = {
						value: JSON.stringify(Object.fromEntries(apsWithDevices))
						,
						type: "connectedDevices"
					};
					console.debug(msg	);
					fetchCallback(self, msg);
				}
			}
			else {
				console.error('Error happened: ' + error);
			}
		});
	};

	
	
	var callbackOpenSession = function(error, response, body){
        if (!error) {
            var info = JSON.parse(JSON.stringify(body));    
			console.debug(info);        
            if (info.success == true){
                config.session_token = info.result.session_token;
                var outputFilename = 'freebox.txt';
		fs.writeFile(outputFilename, JSON.stringify(config, null, 4), function(err) {
		if(err) {
			console.error(err);
		} else {
			msg = {
				value: "Connected",
				type: "connection"
			};
			fetchCallback(self, msg);
		}
		});
            }
        }
        else {
            console.error('Error happened: ' + error);
        }
    };

    this.authorize = function( app_id ,app_name, app_version, device_name){
        config.appId = app_id
        var data = {
            app_id: app_id,
            app_name: app_name,
            app_version: app_version,
            device_name: device_name
        };
        var client = reqjson.createClient(ip);
		try {
			fs.accessSync("freebox.txt", fs.F_OK);			
			config = JSON.parse(fs.readFileSync('freebox.txt', 'utf8'));
			console.debug(config);
			getChallenge(callbackGetChallenge);
		} catch (e) {
			// It isn't accessible
			client.post('/api/v'+config.version+'/login/authorize/', data, authCallback);
		}
    };

    var checkAuthorize = function(track_id, callback){
        var client = reqjson.createClient(ip);
        client.get('/api/v'+config.version+'/login/authorize/'+config.track_id, callback);
    };

    var getChallenge = function(callback){
        var client = reqjson.createClient(ip);
        client.get('/api/v'+config.version+'/login/', callback);
    };


    var openSession = function(callback){        
        var client = reqjson.createClient(ip);
        var message = config.challenge;
        var passphrase = config.app_token;
        var signature = CryptoJS.enc.Hex.stringify(CryptoJS.HmacSHA1( message, passphrase));
        var data = {
            app_id: config.appId,
            password: signature
        };
        client.post('/api/v'+config.version+'/login/session/', data, callbackOpenSession);
    };

	this.getConnectedDevices = function(){
		var client = reqjson.createClient(ip);
        client.headers['X-Fbx-App-Auth'] = config.session_token;
		//retrieve ap list
        client.get('/api/v'+config.version+'/wifi/ap/', apListCallback	);
	};

	this.getConnectionStatus = function(){
		var client = reqjson.createClient(ip);
        client.headers['X-Fbx-App-Auth'] = config.session_token;
        client.get('/api/v'+config.version+'/connection/', connectionStatusCallback	);
	};

    this.getCalls = function(){
        var client = reqjson.createClient(ip);
        client.headers['X-Fbx-App-Auth'] = config.session_token;
        client.get('/api/v'+config.version+'/call/log/', callbackCalls);
    };

	this.getDownloads = function(){
		var client = reqjson.createClient(ip);
        client.headers['X-Fbx-App-Auth'] = config.session_token;
        client.get('/api/v'+config.version+'/downloads/', callbackDownloads);
	};

	this.getSystemInfo = function(){
		var client = reqjson.createClient(ip);
        client.headers['X-Fbx-App-Auth'] = config.session_token;
        client.get('/api/v'+config.version+'/system/', callbackSystemInfo);
	};

	this.onError = function(callback) {
		fetchFailedCallback = callback;
	};

	this.onMsg = function(callback) {
		fetchCallback = callback;
	};
};

module.exports = Freeboxapi;
